runtime/v2/sandbox.go

```diff
@@ -0,0 +1,254 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v2
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/runtime"
+	proto "github.com/containerd/containerd/runtime/v2/task"
+	"github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+)
+
+// SandboxManager is a layer on top of task manager to support sandboxed environments.
+// Container record now have SandboxID field to identify which sandbox the container belongs to.
+// If it's empty, SandboxManager will fallback to TaskManager's implementation to handle container tasks.
+// Otherwise sandbox manager will find proper shim instance and will route query to that shim.
+type SandboxManager struct {
+	tasks     *TaskManager
+	sandboxes *runtime.TaskList
+}
+
+var _ runtime.SandboxRuntime = (*SandboxManager)(nil)
+var _ runtime.PlatformRuntime = (*SandboxManager)(nil)
+
+// NewSandboxManager will create a sandbox manager to control sandboxes on top of the existing v2 task manager.
+func NewSandboxManager(tasks *TaskManager) *SandboxManager {
+	return &SandboxManager{
+		tasks:     tasks,
+		sandboxes: runtime.NewTaskList(),
+	}
+}
+
+// Start will launch a new sandbox instance task.
+func (m *SandboxManager) Start(ctx context.Context, sandboxID string, opts runtime.SandboxOpts) (_ runtime.Process, retErr error) {
+	if _, err := m.getShim(ctx, sandboxID); err == nil {
+		return nil, errors.Wrapf(errdefs.ErrAlreadyExists, "sandbox %s already running", sandboxID)
+	}
+
+	bundle, err := NewBundle(ctx, m.tasks.root, m.tasks.state, sandboxID, opts.Spec.Value)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create bundle")
+	}
+
+	shim, err := m.tasks.startShim(ctx, bundle, opts.RuntimeName, opts.RuntimeOpts, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to start shim")
+	}
+
+	svc, err := shim.Sandbox(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to retrieve sandbox service from shim")
+	}
+
+	if _, err := svc.StartSandbox(ctx, &proto.StartSandboxRequest{
+		SandboxID:  sandboxID,
+		BundlePath: bundle.Path,
```

Comment:
How about adding an parameter to pass on sandbox resource information? The "opts.RuntimeOpts" should be used to configure the shim, and we still lack a mechanism to configure the sandbox resource information.
Take an vm based secure container as an example, "opts.RuntimeOpts" may be used to pass vmm security configurations, and we may need another parameter to configure the initial CPUs/memory size of the vm.

Reply:
yes or maybe a todo to add it later..